File: src/forager/artwork/pipeline.py
"""Art loading for games: header / grid / hero / logo images.

Resolution order (per image): local Steam appcache art (Steam games), the on
disk art cache, the Steam CDN for the resolved ``app_id``, SteamGridDB by app
id, SteamGridDB by search, then the generated placeholders.
"""
from __future__ import annotations
import threading
from pathlib import Path
from PySide6.QtGui import QPixmap

from forager.core.game import Game, Source
from forager.services.icon_provider import load_icon
from forager.services.steamgriddb import (
    fetch_header_bytes_for_steam,
    fetch_banner_bytes_for_steam,
    fetch_grid_bytes_for_steam, fetch_grid_bytes_for_game,
    fetch_header_bytes_for_game,
    fetch_banner_bytes_for_game,
)
from forager.artwork.pixmap_utils import bytes_to_pixmap, scale_crop, scaled
from forager.artwork.placeholder import (
    placeholder_card,
    placeholder_grid,
    register_placeholder_font,
)
from forager.providers.steam.appid import steam_app_id
from forager.artwork.cache import (
    art_cache_dir, banner_cache_dir, cache_key, cached_path,
)
from forager.core.paths import steam_appcache_dir

STEAM_CACHE = steam_appcache_dir()
ART_CACHE = art_cache_dir()
BANNER_CACHE = banner_cache_dir()
STEAM_CDN = "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/{app_id}/{name}"
_CACHE_LOCK = threading.Lock()


def _ensure_cache():
    ART_CACHE.mkdir(parents=True, exist_ok=True)


def _steam_path(game: Game, filename: str) -> Path | None:
    if game.source != Source.STEAM or not game.app_id:
        return None
    p = STEAM_CACHE / game.app_id / filename
    return p if p.is_file() else None


# -- Steam CDN ----------------------------------------------------------

def _steam_cdn_bytes(app_id: str, names: tuple[str, ...]) -> bytes | None:
    from forager.utils.network import http_get

    for name in names:
        url = STEAM_CDN.format(app_id=app_id, name=name)
        try:
            return http_get(url)
        except Exception:
            continue
    return None


def _steam_cdn_grid_bytes(app_id: str) -> bytes | None:
    return _steam_cdn_bytes(app_id, ("library_600x900.jpg", "library_600x900_2x.jpg"))


# -- on-disk caches -----------------------------------------------------

def _image_ext(data: bytes) -> str:
    return ".jpg" if data[:3] == b"\xff\xd8" else ".png"


def _cached_header_path(game: Game) -> Path | None:
    return cached_path(ART_CACHE, "header", cache_key(game.app_id or game.name))


def _cached_grid_path(game: Game) -> Path | None:
    return cached_path(ART_CACHE, "grid", cache_key(game.app_id or game.name))


def _cached_hero_path(game: Game) -> Path | None:
    return cached_path(BANNER_CACHE, "hero", cache_key(game.app_id or game.name))


# -- header -------------------------------------------------------------

def load_header_bytes(game: Game, allow_network: bool = True) -> bytes | None:
    local = _steam_path(game, "header.jpg")
    if local is not None:
        return local.read_bytes()

    cached = _cached_header_path(game)
    if cached is not None:
        return cached.read_bytes()

    if not allow_network:
        return None

    data = None
    app_id = steam_app_id(game)
    if app_id:
        data = _steam_cdn_bytes(app_id, ("header.jpg",))
    if data is None and app_id:
        data = fetch_header_bytes_for_steam(app_id)
    if data is None:
        data = fetch_header_bytes_for_game(game)
    if data:
        _ensure_cache()
        key = cache_key(game.app_id or game.name)
        with _CACHE_LOCK:
            (ART_CACHE / f"header_{key}.png").write_bytes(data)
    return data


def load_header(game: Game, allow_network: bool = True) -> QPixmap | None:
    data = load_header_bytes(game, allow_network)
    return bytes_to_pixmap(data) if data else None


# -- grid ---------------------------------------------------------------

def load_grid_bytes(game: Game, allow_network: bool = True) -> bytes | None:
    if game.source == Source.STEAM and game.app_id:
        for name in ("library_600x900.jpg", "header.jpg"):
            local = _steam_path(game, name)
            if local is not None:
                return local.read_bytes()

    cached = _cached_grid_path(game)
    if cached is not None:
        return cached.read_bytes()

    if not allow_network:
        return None

    data = None
    app_id = steam_app_id(game)
    if app_id:
        data = _steam_cdn_grid_bytes(app_id)
    if data is None and app_id:
        data = fetch_grid_bytes_for_steam(app_id)
    if data is None:
        data = fetch_grid_bytes_for_game(game)
    if data:
        _ensure_cache()
        key = cache_key(game.app_id or game.name)
        with _CACHE_LOCK:
            (ART_CACHE / f"grid_{key}{_image_ext(data)}").write_bytes(data)
    return data


def load_grid(game: Game, allow_network: bool = True) -> QPixmap | None:
    data = load_grid_bytes(game, allow_network)
    return bytes_to_pixmap(data) if data else None


# -- hero ---------------------------------------------------------------

def load_hero_bytes(game: Game, allow_network: bool = True) -> bytes | None:
    for name in ("library_hero.jpg", "library_hero_blur.jpg"):
        local = _steam_path(game, name)
        if local is not None:
            return local.read_bytes()

    cached = _cached_hero_path(game)
    if cached is not None:
        return cached.read_bytes()

    if not allow_network:
        return None

    data = None
    app_id = steam_app_id(game)
    if app_id:
        data = _steam_cdn_bytes(app_id, ("library_hero.jpg", "library_hero_blur.jpg"))
    if data is None and app_id:
        data = fetch_banner_bytes_for_steam(app_id)
    if data is None:
        data = fetch_banner_bytes_for_game(game)
    if data:
        BANNER_CACHE.mkdir(parents=True, exist_ok=True)
        key = cache_key(game.app_id or game.name)
        with _CACHE_LOCK:
            (BANNER_CACHE / f"hero_{key}{_image_ext(data)}").write_bytes(data)
        return data
    return load_header_bytes(game, allow_network)


def load_hero(game: Game, allow_network: bool = True) -> QPixmap | None:
    for name in ("library_hero.jpg", "library_hero_blur.jpg", "header.jpg"):
        local = _steam_path(game, name)
        if local is not None:
            pix = QPixmap(str(local))
            if not pix.isNull():
                return pix
    cached = _cached_hero_path(game)
    if cached is not None:
        pix = QPixmap(str(cached))
        if not pix.isNull():
            return pix
    data = load_hero_bytes(game, allow_network)
    return bytes_to_pixmap(data) if data else None


# -- logo ---------------------------------------------------------------

def load_logo(game: Game) -> QPixmap | None:
    local = _steam_path(game, "logo.png")
    if local is not None:
        pix = QPixmap(str(local))
        if not pix.isNull():
            return pix
    return None

File: src/artwork/Pipeline.cpp
#include "artwork/Pipeline.h"
#include "artwork/PipelineInternal.h"
#include "artwork/Cache.h"
#include "artwork/Placeholder.h"
#include "artwork/PixmapUtils.h"
#include "core/Paths.h"
#include "core/Game.h"
#include "providers/steam/SteamAppId.h"
#include "services/SteamGridDB.h"
#include "utils/Network.h"
#include "utils/Filesystem.h"

#include <QPixmap>
#include <QImage>
#include <QBuffer>
#include <QFile>
#include <algorithm>

using art::detail::fetchAssetByGameId;
using art::detail::fetchSteamCDN;
using art::detail::fetchSGDBBytes;
using art::detail::fromDiskCache;
using art::detail::fromSteamAppcache;
using art::detail::isSteamLocal;
using art::detail::loadPixmapBytes;
using art::detail::loadPixmapFile;
using art::detail::looksLikeImage;
using art::detail::saveToCache;
using art::detail::sgdbGameIdForSteamApp;

QPixmap art::detail::loadPixmapFile(const QString& path) {
    QPixmap p;
    if (!path.isEmpty()) p.load(path);
    return p;
}

QPixmap art::detail::loadPixmapBytes(const QByteArray& data) {
    QPixmap p;
    if (!data.isEmpty()) p.loadFromData(data);
    return p;
}

bool art::detail::looksLikeImage(const QByteArray& data) {
    if (data.size() < 4) return false;
    if (data[0] == '\x89' && data[1] == 'P') return true;
    if (static_cast<unsigned char>(data[0]) == 0xFF && static_cast<unsigned char>(data[1]) == 0xD8) return true;
    if (data.startsWith("RIFF") && data.mid(8, 4) == "WEBP") return true;
    return false;
}

bool art::detail::saveToCache(const QString& dir, const QString& key, const QByteArray& data) {
    if (!looksLikeImage(data)) return false;
    QString ext = (data[0] == '\x89') ? QStringLiteral(".png") : QStringLiteral(".jpg");
    return fs::writeBytes(dir + "/" + key + ext, data);
}

// --- Disk cache ---

QPixmap art::detail::fromDiskCache(const QString& cacheDir, const QString& key) {
    return loadPixmapFile(artcache::cachedPath(cacheDir, key));
}

// --- Unified load pipeline ---

static QPixmap loadArt(const Game& game, bool allowNetwork, bool isGrid) {
    QString key = artcache::cacheKey(game.identifier());
    static const char* gridFiles[] = {"library_600x900.jpg", "library_600x900_2x.jpg", "header.jpg"};
    static const char* heroFiles[] = {"library_hero.jpg", "library_hero_blur.jpg", "header.jpg"};
    const char** cdnFiles = isGrid ? gridFiles : heroFiles;
    int cdnCount = isGrid ? 3 : 3;
    const QString& cacheDir = isGrid ? paths::artCacheDir() : paths::bannerCacheDir();
    const QString cachePrefix = isGrid ? QStringLiteral("grid_") : QStringLiteral("hero_");
    const QString cacheKey = cachePrefix + key;

    // 1. Local steam appcache (Steam only)
    QPixmap p;
    for (int i = 0; i < cdnCount && p.isNull(); ++i)
        p = fromSteamAppcache(game, cdnFiles[i]);
    if (!p.isNull()) return p;

    // 2. Disk cache
    p = fromDiskCache(cacheDir, cacheKey);
    if (!p.isNull()) return p;

    if (!allowNetwork) {
        if (isGrid)
            return placeholder::placeholderGrid(game.name(), 165, 248);
        return p;
    }

    // 3. Steam CDN via resolved app id
    QString appId = SteamAppId::resolve(game).value_or(QString());
    QByteArray cdn;
    for (int i = 0; i < cdnCount && cdn.isEmpty(); ++i)
        cdn = fetchSteamCDN(appId, cdnFiles[i]);
    if (!cdn.isEmpty()) {
        saveToCache(cacheDir, cacheKey, cdn);
        p = loadPixmapBytes(cdn);
        if (!p.isNull()) return p;
    }

    // 4. SteamGridDB
    QByteArray sgdbData = fetchSGDBBytes(game, isGrid);
    if (!sgdbData.isEmpty()) {
        saveToCache(cacheDir, cacheKey, sgdbData);
        p = loadPixmapBytes(sgdbData);
        if (!p.isNull()) return p;
    }

    // 5. Hero falls back to header
    if (!isGrid && !appId.isEmpty()) {
        QByteArray headerData = fetchSteamCDN(appId, "header.jpg");
        if (!headerData.isEmpty()) {
            saveToCache(cacheDir, cacheKey, headerData);
            p = loadPixmapBytes(headerData);
            return p;
        }
    }

    if (isGrid)
        return placeholder::placeholderGrid(game.name(), 165, 248);
    return p;
}

// --- Public API ---

QPixmap art::loadGrid(const Game& game, bool allowNetwork) {
    return loadArt(game, allowNetwork, true);
}

QPixmap art::loadHero(const Game& game, bool allowNetwork) {
    return loadArt(game, allowNetwork, false);
}

QPixmap art::loadIcon(const Game& game, bool allowNetwork) {
    const QString& iconDir = paths::iconCacheDir();
    QString key = artcache::cacheKey(game.identifier());
    QString cached = artcache::cachedPath(iconDir, key);
    if (!cached.isEmpty()) {
        QPixmap p = loadPixmapFile(cached);
        if (!p.isNull()) return p.scaled(48, 48, Qt::KeepAspectRatio, Qt::SmoothTransformation);
    }

    // Local steam logo (square crop)
    if (isSteamLocal(game)) {
        QString logoPath = paths::steamAppcacheDir() + "/" + game.appId() + "/logo.png";
        QImage img(logoPath);
        if (!img.isNull()) {
            int size = std::min(img.width(), img.height());
            if (size > 0) {
                int ox = (img.width() - size) / 2;
                int oy = (img.height() - size) / 2;
                QImage square = img.copy(ox, oy, size, size);
                QPixmap pix = QPixmap::fromImage(square).scaled(
                    48, 48, Qt::KeepAspectRatio, Qt::SmoothTransformation);
                if (!pix.isNull()) {
                    QByteArray png;
                    QBuffer buf(&png);
                    buf.open(QIODevice::WriteOnly);
                    if (pix.save(&buf, "PNG"))
                        fs::writeBytes(iconDir + "/" + key + ".png", png);
                    return pix;
                }
            }
        }
    }

    // Local icon files next to the game
    if (!game.path().isEmpty()) {
        for (const char* name : {"icon.png", "icon.ico", "icon.svg", "Icon.png", "Icon.ico"}) {
            QString candidate = game.path() + "/" + name;
            if (QFile::exists(candidate)) {
                QPixmap pix(candidate);
                if (!pix.isNull())
                    return pix.scaled(48, 48, Qt::KeepAspectRatio, Qt::SmoothTransformation);
            }
        }
        QString mcIcon = game.path() + "/.minecraft/icon.png";
        if (QFile::exists(mcIcon)) {
            QPixmap pix(mcIcon);
            if (!pix.isNull())
                return pix.scaled(48, 48, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        }
    }

    if (!allowNetwork) return {};

    auto& sgdb = SteamGridDB::instance();
    if (sgdb.isConfigured()) {
        QByteArray data;
        QString appId = isSteamLocal(game) ? game.appId()
                                           : SteamAppId::resolve(game).value_or(QString());
        if (game.source() == Source::Steam && !appId.isEmpty()) {
            QString sgdbId = sgdbGameIdForSteamApp(appId);
            if (!sgdbId.isEmpty())
                data = net::httpGet(
                    QStringLiteral("https://www.steamgriddb.com/api/v2/icons/game/%1").arg(sgdbId),
                    15000, net::bearerHeaders(sgdb.token()));
        }
        if (data.isEmpty() && game.source() != Source::Steam)
            data = fetchAssetByGameId(game, "icons", false);
        if (!data.isEmpty() && looksLikeImage(data)) {
            fs::writeBytes(iconDir + "/" + key + ".png", data);
            return loadPixmapBytes(data).scaled(48, 48, Qt::KeepAspectRatio, Qt::SmoothTransformation);
        }
    }
    return {};
}

QByteArray art::loadGridBytes(const Game& game) {
    return pixmap::toJpeg(loadGrid(game, false));
}

QByteArray art::loadHeroBytes(const Game& game) {
    return pixmap::toJpeg(loadHero(game, false));
}

void art::registerPlaceholderFont() {}
